doc/develop/west/manifest.rst

```diff
@@ -258,6 +258,9 @@ next.
 
        The default ``revision`` is ``master`` if not otherwise specified.
 
+       Using HEAD~0 as the ``revision`` will cause west to keep the current
```

Comment:
Seems like a verbose way of saying `HEAD`. Why do you need the `~0`?

Reply:
because using `HEAD` would result in `west` fetching remote `HEAD`, like this:
```
=== updating zephyr (zephyr):
--- zephyr: fetching, need revision HEAD
....
From https://github.com/zephyrproject-rtos/zephyr
 * branch                  HEAD       -> FETCH_HEAD
```

but in this particular case, users want to work with Zephyr, just as if it was using Zephyr as manifest repo, that is `git fetch`, `git checkout`, `git bisect`, etc. are done directly in Zephyr repo, without `west` interfering with that repo when doing `west update`.

See it as a filtering done without patching `zephyr` west.yml.

That's where `HEAD~0` comes in handy, cause that tells `west` (`git`) to keep using the local `HEAD` and not fetching remote `HEAD` ;-)

Reply:
OK, that's a cute hack :).

I'm OK with keeping this part of the patch, but I'd like to see a test case added to `west` to ensure this first. That way I won't accidentally break it later if people are relying on it.

Reply:
Where should the test go (which test file/function)?

Reply:
> which test file/function

Add a new `def test_update_head_0` in https://github.com/zephyrproject-rtos/west/blob/main/tests/test_project.py which has a workspace that uses this trick. The test must verify that 1. west doesn't fetch from the remote (easy way to do this is with an invalid URL), and 2. the project revision in the workspace is unchanged.

Reply:
@mbolivar-nordic https://github.com/zephyrproject-rtos/west/pull/605

I reduced the scope of this PR as well.

Reply:
> OK, that's a cute hack :).

Hack => a minimum of documentation and labelling at such.

Can you please add at least a vague clue about why `west` treats `HEAD` and `HEAD~0` differently? Calling this "unexpected" is an understatement.

Right now the total lack of explanation can very easily be misinterpreted as something `git` treats differently. Or worse.

Reply:
Nice hack :) Please use double backticks for `HEAD~0` and add some extra doc as @marc-hb mentions

Reply:
> Can you please add at least a vague clue about why `west` treats `HEAD` and `HEAD~0` differently? Calling this "unexpected" is an understatement.

the reason as to why `west` treats `HEAD` and `HEAD~0` differently comes down to what `git` reports.
`west` invokes:
```
$ git rev-parse --verify --symbolic-full-name 'HEAD'
HEAD
$
```

and
```
$ git rev-parse --verify --symbolic-full-name 'HEAD~0'
$
```

The `HEAD` result is interpreted as `other` and an empty result is interpreted as a commit, ref:
https://github.com/zephyrproject-rtos/west/blob/ed01ab5fa1a918da3661733839be90d31ea94809/src/west/app/project.py#L1578-L1581

A `commit` will not result in a new fetch, but `other` will:
https://github.com/zephyrproject-rtos/west/blob/ed01ab5fa1a918da3661733839be90d31ea94809/src/west/app/project.py#L1246-L1247
```
        if self.fs == 'always' or _rev_type(project) not in ('tag', 'commit'):
            self.fetch(project, stats, take_stats)
```

Reply:
`--symbolic-full-name`... wow!

I assumed `west`'s parsing was making a different between HEAD and HEAD~0 but I clearly underestimated what one can find in the darkest corners of git.

https://git-scm.com/docs/git-rev-parse
```

       --symbolic
           Usually the object names are output in SHA-1 form (with possible ^ prefix);
           this option makes them output in a form as close
           to the original input as possible.

       --symbolic-full-name
           This is similar to --symbolic, but it omits input that are not refs (i.e.
           branch or tag names; or more explicitly
           disambiguating "heads/master" form, when you want to name the "master" 
            branch when there is an unfortunately named tag
           "master"), and show them as full refnames (e.g. "refs/heads/master").
```           
         

I think the requested "vague clue" should more or less quote git's documentation, probably something like "this works because HEAD is a git reference while HEAD~0 is not"



For anyone curious this is what happens when the repo is NOT in "detached HEAD" mode, in other words when there is a `current_branch`.
```
git rev-parse --symbolic HEAD
HEAD

git rev-parse --symbolic HEAD~0
HEAD~0

git rev-parse --symbolic HEAD~1
HEAD~1

git rev-parse --symbolic-full-name HEAD
refs/heads/current_branch

git rev-parse --symbolic-full-name HEAD~0
<empty>

git rev-parse --symbolic-full-name HEAD~1
<empty>
```

Good luck making sense of that!

Reply:
> I think the requested "vague clue" should more or less quote git's documentation, probably something like "this works because HEAD is a git reference while HEAD~0 is not"

How about,

> In git, `HEAD` is a reference, whereas `HEAD~<n>` is a specific commit. `west` fetches references, such as `refs/heads/main` or `HEAD`, and commits not available locally, but will not fetch commits if they are already available.
`HEAD~0` is resolved to a specific commit that is locally available, and therefore `west` will simply checkout the locally available commit, identified by `HEAD~0`.

Reply:
> HEAD~\<n> is a specific commit

I think this is inaccurate and missing the critical point which is: "not a reference".

> HEAD~0 is resolved to a specific commit

More accurate.

https://git-scm.com/docs/git-rev-parse#_specifying_revisions calls `HEAD~0` a "revision" where a revision "typically names a commit object".

So I would change only your first sentence, the rest looks great to me:

"In git, HEAD is a reference, whereas HEAD~\<n> is a valid revision but not a reference".

Reply:
> "In git, HEAD is a reference, whereas HEAD~<n> is a valid revision but not a reference".

thanks for the improvement.
Completely agree, and this is more precise in what I wanted to describe.

Reply:
Thanks! I added the text as a footnote, to not clutter the table. @tejlmand @marc-hb (@mbolivar-nordic) please take another look.